### ISSUE TITLE
disable recursive sub directory for completion doesn't work

### ISSUE THREAD

**Issue by atjshop** [romkatv/zsh4humans#115]
I tried to disable recursive sub directory for completion. 
According to https://github.com/romkatv/zsh4humans/issues/107
I updated z4h by z4h update, then add below line to .zshrc
zstyle ':z4h:fzf-complete' recurse-dirs 'no'
the restart shell by exec zsh
however the subdirectories still show up in the tab completion for cd command.
how do I troubleshooting this?

I think the recursive is there for a reason, I just don't used to that. I see the benefit of this recursive when I need to go do deep subdirectory. But most of time (for me) is go to one of immediate subdirectory. Supposed that there is two subdirectories in the current directory that start with "ab", when I run cd ab<tab>, I'd expect it completes with these two subdirectories and press <tab> to cycle through them and press enter to select one. But in z4h, when I run cd ab<tab>, it pops up a full screen of completion candidates above the command prompt, which includes all of the subdirectories too, then has to either navigate to the desired directory or continue type to narrow down the selection, then press enter twice to go to the directory. Seems not efficient to me. Am I using it wrong way?

**Comment by romkatv** [comment]
> I tried to disable recursive sub directory for completion.
> According to #107
> I updated z4h by z4h update, then add below line to .zshrc
> zstyle ':z4h:fzf-complete' recurse-dirs 'no'
> the restart shell by exec zsh
> however the subdirectories still show up in the tab completion for cd command.

I've tried it and it indeed doesn't work. Sorry about that and thanks for reporting this issue. Fixed in https://github.com/romkatv/zsh4humans/commit/85cb8c566544af901c91f3b047e482d4bc2c40b2.

> I think the recursive is there for a reason, I just don't used to that.

Recursive file completions are really good but to use them efficiently you'll need to change your habits. If you use shell a lot, it's worth going through the trouble adaptation period. If you don't, then it makes sense to disable recursive completions.

What I mean by changing habits is that instead of attempting to complete everything left-to-right (`rm foo<TAB>/bar<TAB>/baz<TAB>`) you complete it based on the most salient terms (`rm <TAB>` and then perhaps "bar"; if there are too many matches, I would add "$" after "bar"; if there are still too many matches, I would type space followed by more terms -- perhaps "foo"). Sort of like using an internet catalog on Yahoo! in the 90s vs using Google today.

Specifically for `cd` there are specialized tools in zsh4humans that are often more convenient. If you want to `cd` into a subdirectory, press <kbd>Alt+Down</kbd> (<kbd>Shift+Down</kbd> on Mac). There is also <kbd>Alt+Left</kbd>/<kbd>Alt+Right</kbd> to go back/forward (like  in a web browser) and <kbd>Alt+Up</kbd> to go to the parent dir. I've recent also added <kbd>Alt+R</kbd> for persistent history of directories.

That said, you don't have to use these features. zsh4humans can be useful even without them.

**Comment by romkatv** [comment]
> Thanks for the super fast response. I tried z4h update but seems not getting the new commit. How can I get your new update?

`z4h update` will pick up the new commit as long as you are using v5. You can check which version you are on with `print -r -- $Z4H_URL`.

> As for the shift+down shortcut, it is also showing recursive subdirectory.

This is intended.

> Does the new setting apply to this shortcut too?

No. The purpose of the new option is to allow people with strong muscle memory to use zsh4humans without relearning habits. No one has muscle memory for shift+down. I could add a similar option for it if the need arises.

In general, I don't intend to build a zsh framework that is so flexible that it can be configured to satisfy any wish. My goal is more modest -- to build a great shell environment.

**Comment by romkatv** [comment]
> > Thanks for the super fast response. I tried z4h update but seems not getting the new commit. How can I get your new update?
> 
> `z4h update` will pick up the new commit as long as you are using v5. You can check which version you are on with `print -r -- $Z4H_URL`.

Oh, that's embarrassing. There was another bug that prevented the new option from working properly. Should be fixed now.

**Comment by atjshop** [comment]
I am using v5:
❯ print -r -- $Z4H_URL
https://raw.githubusercontent.com/romkatv/zsh4humans/v5

and have this config in the .zshrc
❯ rg 'recurse-dirs' ~/.zshrc
29:zstyle ':z4h:fzf-complete' recurse-dirs 'no'

I did a z4h update, however recursive subdirectories still showing up. Tried restart zsh/iterm2/reboot, still same :(

**Comment by romkatv** [comment]
> I did a z4h update, however recursive subdirectories still showing up.

They show up when you do *what*?

Please post the output of the following command:

```zsh
() {
  local val
  zstyle -s :z4h:fzf-complete recurse-dirs val
  typeset -p val
  command grep recurse -- $Z4H/zsh4humans/fn/-z4h-comp-files
}
```

**Comment by atjshop** [comment]
> They show up when you do _what_?
Recursive suddirectory show up when I do 'cd' command using autocompletion. In below I create three directories 'abc', 'abc/subabc' and 'abd', then try 'cd ab<tab>', I expect only 'abc' and 'abd' appear on the completion list. However the subdirectory abc/subabc also appeared in the list.

❯ mkdir -p abc abc/subabc abd

  abc/subabc
  abd
> abc
  3/3 (0)
> ^ab
❯ cd ab..

Below is the output for the command:
❯ () {
 local val
 zstyle -s :z4h:fzf-complete recurse-dirs val
 typeset -p val
 command grep recurse -- $Z4H/zsh4humans/fn/-z4h-comp-files
}
typeset val=no
  local -i recurse
    # Set curcontext and dot_glob in case the value of recurse depends on them (via `zstyle -e`).
    zstyle -T :z4h:${WIDGET#z4h-} recurse-dirs && recurse=1
  local -i height=$(( ! (recurse && $#dirs) && 100 * ($#_z4h_words + 4) < pct * LINES ? $#_z4h_words + 4 : pct * LINES / 100 ))
        if (( recurse )); then

**Comment by romkatv** [comment]
Fascinating. I cannot reproduce this. Could you post the output of `functions -- -z4h-comp-files`? After that, run `functions -t -- -z4h-comp-files`, type `ab<TAB>` (with the same directory structure as before) and post all output (there will be about two screenfuls of it).

**Comment by atjshop** [comment]
Here is the output of the function:
====
❯ functions -- -z4h-comp-files
-z4h-comp-files () {
        () {
                eval "$_z4h_opt"
                local -i dot_glob list_types
                [[ $1 == on ]] && dot_glob=1
                [[ $2 == on ]] && list_types=1
                if (( _z4h_only_dirs ))
                then
                        local dirs=($_z4h_path_prefix${^${(Q)_z4h_words}}/*(D-/Y1N:h:t))
                else
                        local dirs=($_z4h_path_prefix${^${(Q)_z4h_words}}/*(DY1N:h:t))
                fi
                -z4h-set-list-colors "$_z4h_curcontext" "$list_types"
                local -i list_colors=$((!$?))
                autoload +X -Uz -- -z4h-present-files -z4h-cursor-show -z4h-find -z4h-fzf
                local -i pct=60
                (( _z4h_can_save_restore_screen )) && pct=100
                local -i recurse
                () {
                        emulate -L zsh
                        (( dot_glob )) && setopt dot_glob
                        local curcontext=$_z4h_curcontext
                        zstyle -T :z4h:${WIDGET#z4h-} recurse-dirs && recurse=1
                }
                local -i height=$(( ! (recurse && $#dirs) && 100 * ($#_z4h_words + 4) < pct * LINES ? $#_z4h_words + 4 : pct * LINES / 100 ))
                (( height >= 6 )) || (( height = 6 ))
                (( height <= LINES - 1 )) || (( height = LINES - 1 ))
                local opts=(--query=${_z4h_word_prefix:+"^$_z4h_word_prefix"} --color=hl:201,hl+:201 --with-nth=2 --delimiter='\000' --ansi --exact --no-mouse --tiebreak=length,begin,index --multi --cycle --border=horizontal)
                () {
                        emulate -L zsh
                        (( dot_glob )) && setopt dot_glob
                        local -a bin
                        zstyle -a :z4h:${WIDGET#z4h-} find-command bin
                        if (( ! $#bin )) && (( $+commands[bfs] ))
                        then
                                opts+=(--no-sort)
                        fi
                }
                local cursor_y cursor_x
                -z4h-get-cursor-pos || return
                if (( _z4h_can_save_restore_screen ))
                then
                        opts+=(--no-clear)
                        if {
                                        (( height <= cursor_y - 1 )) && zstyle -T :z4h: prompt-at-bottom
                                } || (( cursor_y - 1 > LINES - cursor_y && cursor_y - 1 > 6 )) && {
                                        (( height > LINES - cursor_y )) || zstyle -T :z4h: prompt-at-bottom
                                }
                        then
                                (( height <= cursor_y - 1 )) || (( height = cursor_y - 1 ))
                                local move=$'\e[0m\e['$((cursor_y-height))';1H'
                                opts+=(--layout=default)
                        elif (( LINES - cursor_y > 6 ))
                        then
                                (( height <= LINES - cursor_y )) || (( height = LINES - cursor_y ))
                                local move=$'\e[0m\n\r'
                                opts+=(--layout=reverse)
                        else
                                local -i extra=$((height - LINES + cursor_y))
                                print -rn -- ${(pl:$height::\n:)} > $TTY || return
                                (( cursor_y += LINES - cursor_y - height ))
          #       reply=(...)
          #     fi
          #   }
          local curcontext=$_z4h_curcontext
          -z4h-find $dot_glob $_z4h_only_dirs $dirs | command cut -b3-
        fi
      } | {
        -z4h-present-files $list_colors $list_types 0
      } | {
        print -rn -- $move >$TTY
        -z4h-cursor-show
        2>$TTY -z4h-fzf $opts
      }
    )"
                } always {
                        -z4h-cursor-hide
                        if (( _z4h_can_save_restore_screen ))
                        then
                                -z4h-restore-screen
                                print -rn -- $'\e[0m\e['$cursor_y';'$cursor_x'H'
                        else
                                builtin echoti cuu 1
                                (( cursor_x > 1 )) && builtin echoti cuf $((cursor_x-1))
                        fi
                }
                [[ -n $choice ]] || return
                choice=("${(@f)choice}")
                typeset -g _z4h_reply=(0 ${${choice:1}%$'\0'*})
                [[ -z $choice[1] ]] || _z4h_reply[1]=1
        } "${options[dot_glob]}" "${options[list_types]}"
}
======

below is the output of ab<tab>:
=====
❯ cd ab+-z4h-comp-files:3> '(anon)' on off
+(anon):1> eval $'emulate -L zsh &&\n  setopt typeset_silent pipe_fail extended_glob prompt_percent no_prompt_subst &&\n  setopt no_prompt_bang no_bg_nice no_aliases'
+(eval):1> emulate -L zsh
+(eval):2> setopt typeset_silent pipe_fail extended_glob prompt_percent no_prompt_subst
+(eval):3> setopt no_prompt_bang no_bg_nice no_aliases
+(anon):3> local -i dot_glob list_types
+(anon):4> [[ on == on ]]
+(anon):4> dot_glob=1
+(anon):5> [[ off == on ]]
+(anon):7> ((  _z4h_only_dirs  ))
+(anon):8> local dirs=( abc )
+(anon):13> -z4h-set-list-colors :complete:cd: 0
+-z4h-set-list-colors:3> eval $'emulate -L zsh &&\n  setopt typeset_silent pipe_fail extended_glob prompt_percent no_prompt_subst &&\n  setopt no_prompt_bang no_bg_nice no_aliases'
+(eval):1> emulate -L zsh
+(eval):2> setopt typeset_silent pipe_fail extended_glob prompt_percent no_prompt_subst
+(eval):3> setopt no_prompt_bang no_bg_nice no_aliases
+-z4h-set-list-colors:4> setopt octalzeroes
+-z4h-set-list-colors:6> local curcontext=:complete:cd:
+-z4h-set-list-colors:7> local list_types=0
+-z4h-set-list-colors:9> local -a list_colors
+-z4h-set-list-colors:10> zstyle -a :completion::complete:cd::default list-colors list_colors
+zstyle:2> [[ 4 == 3 ]]
+zstyle:5> zstyle -a :completion::complete:cd::default list-colors list_colors
+-z4h-set-list-colors:11> ((  19  ))
+-z4h-set-list-colors:13> [[ $'0fi=00\C-@mi=00\C-@mh=00\C-@ln=01;36\C-@or=01;31\C-@di=01;34\C-@ow=04;01;34\C-@st=34\C-@tw=04;34\C-@pi=01;33\C-@so=01;33\C-@do=01;33\C-@bd=01;33\C-@cd=01;33\C-@su=01;35\C-@sg=01;35\C-@ca=01;35\C-@ex=01;32\C-@no=38;5;248' != 0fi=00mi=00mh=00ln=01;36or=01;31di=01;34ow=04;01;34st=34tw=04;34pi=01;33so=01;33do=01;33bd=01;33cd=01;33su=01;35sg=01;35ca=01;35ex=01;32no=38;5;248 ]]
+-z4h-set-list-colors:57> return 0
+(anon):14> local -i list_colors=1
+(anon):16> autoload +X -Uz -- -z4h-present-files -z4h-cursor-show -z4h-find -z4h-fzf
+(anon):18> local -i pct=60
+(anon):19> ((  _z4h_can_save_restore_screen  ))
+(anon):19> pct=100
+(anon):21> local -i recurse
+(anon):22> '(anon)'
+(anon):1> emulate -L zsh
+(anon):4> ((  dot_glob  ))
+(anon):4> setopt dot_glob
+(anon):5> local curcontext=:complete:cd:
+(anon):6> zstyle -T :z4h:fzf-completion recurse-dirs
+zstyle:2> [[ 3 == 3 && recurse-dirs == list-grouped ]]
+zstyle:5> zstyle -T :z4h:fzf-completion recurse-dirs
+(anon):6> recurse=1
+(anon):31> local -i height=30
+(anon):33> ((  height >= 6  ))
+(anon):34> ((  height <= LINES - 1  ))
+(anon):34> ((  height = LINES - 1  ))
+(anon):36> local opts=( '--query=^ab' '--color=hl:201,hl+:201' '--with-nth=2' '--delimiter=\000' --ansi --exact --no-mouse '--tiebreak=length,begin,index' --multi --cycle '--border=horizontal' )
+(anon):50> '(anon)'
+(anon):1> emulate -L zsh
+(anon):4> ((  dot_glob  ))
+(anon):4> setopt dot_glob
+(anon):5> local -a bin
+(anon):6> zstyle -a :z4h:fzf-completion find-command bin
+zstyle:2> [[ 4 == 3 ]]
+zstyle:5> zstyle -a :z4h:fzf-completion find-command bin
+(anon):7> ((  ! 0  ))
+(anon):7> ((  0  ))
+(anon):62> local cursor_y cursor_x
+(anon):63> -z4h-get-cursor-pos
+-z4h-get-cursor-pos:3> local esc
+-z4h-get-cursor-pos:4> IFS='[;' +-z4h-get-cursor-pos:4> read -s -d R $'esc?\C-[[6n' cursor_y cursor_x
+(anon):65> ((  _z4h_can_save_restore_screen  ))
+(anon):66> opts+=( --no-clear )
+(anon):67> ((  height <= cursor_y - 1  ))
+(anon):67> zstyle -T :z4h: prompt-at-bottom
+zstyle:2> [[ 3 == 3 && prompt-at-bottom == list-grouped ]]
+zstyle:5> zstyle -T :z4h: prompt-at-bottom
+(anon):69> ((  height > LINES - cursor_y  ))
+(anon):70> ((  height <= cursor_y - 1  ))
+(anon):71> local move=$'\C-[[0m\C-[[1;1H'
+(anon):72> opts+=( '--layout=default' )
+(anon):84> local _z4h_saved_screen
+(anon):85> -z4h-save-screen
+-z4h-save-screen:3> eval $'emulate -L zsh &&\n  setopt typeset_silent pipe_fail extended_glob prompt_percent no_prompt_subst &&\n  setopt no_prompt_bang no_bg_nice no_aliases'
+(eval):1> emulate -L zsh
+(eval):2> setopt typeset_silent pipe_fail extended_glob prompt_percent no_prompt_subst
+(eval):3> setopt no_prompt_bang no_bg_nice no_aliases
+-z4h-save-screen:5> [[ -n /tmp/z4h-tmux-502-xterm-256color-1622236387.4660720825,3143,10 ]]
+-z4h-save-screen:6> local pane
+-z4h-save-screen:7> pane=+-z4h-save-screen:7> TMUX=/tmp/z4h-tmux-502-xterm-256color-1622236387.4660720825,3143,10 TMUX_PANE='' /Users/hcheng/.cache/zsh4humans/v5/tmux/bin/tmux capture-pane -p -e
+-z4h-save-screen:7> pane=+-z4h-save-screen:7> print -n x
+-z4h-save-screen:7> pane=$'+zstyle:2> [[ 4 == 3 ]]\n+zstyle:5> zstyle -a :z4h:fzf-completion find-command bin\n+(anon):7> ((  ! 0  ))\n+(anon):7> ((  0  ))\n+(anon):62> local cursor_y cursor_x\n+(anon):63> -z4h-get-cursor-pos\n+-z4h-get-cursor-pos:3> local esc\n+-z4h-get-cursor-+-z4h-fzf:3> eval $'emulate -L zsh &&\n  setopt typeset_silent pipe_fail extended_glob prompt_percent no_prompt_subst &&\n  setopt no_prompt_bang no_bg_nice no_aliases'lear )\n+(anon):67> ((  height <= cursor_y - 1  ))\n+(anon):67> zstyle -T :z4h: prompt-at-bottom\n+zstyle:2> [[ +(eval):1> emulate -L zshm == list-grouped ]]\n+zstyle:5> zstyle -T :z4h: prompt-at-bottom\n+(anon):69> ((  height > LINES - cursor_y  ))\n++(eval):2> setopt typeset_silent pipe_fail extended_glob prompt_percent no_prompt_subst\'\n+(anon):72> opts+=( \'--layout=default\' )\n+(ano+(eval):3> setopt no_prompt_bang no_bg_nice no_aliasesscreen\n+-z4h-save-screen:3> eval $\'emulate -L zsh &&\\n  setopt typeset_silent pipe_+-z4h-fzf:5> local widget=fzf-completionmpt_subst &&\\n  setopt n\no_prompt_bang no_bg_nice no_aliases\'\n+(eval):1> emulate -L zsh\n+(eval)+-z4h-fzf:7> local -a cmd flags bindingsnded_glob prompt_percent no_prompt_subst\n+(eval):3> setopt no_prompt_bang no_bg_nice no_aliases\n+-+-z4h-fzf:8> zstyle -a :z4h:fzf-completion fzf-command cmd1622236387.4660720825,3143,10 ]]\n+-z4h-save-screen:6> local pane\n+-z4h-save-scre+zstyle:2> [[ 4 == 3 ]]creen:7> TMUX=/tmp/z4h-tmux-502-xterm-256color-1622236387.4660720825,3143,10 TMUX_PANE=\'\' /Users/hchen\ng/.cache/zs+zstyle:5> zstyle -a :z4h:fzf-completion fzf-command cmd
+-z4h-fzf:8> cmd=( /Users/hcheng/.cache/zsh4humans/v5/fzf/bin/fzf ) tyle:2> [[ 4 == 3 ]]\C-[[49m\C-[[K\C-M\n+zstyle:5> zstyle -a :z4h:fzf-co+-z4h-fzf:9> zstyle -a :z4h:fzf-completion fzf-flags flags  ! 0  ))\C-[[49m\C-[[K\C-M\n+(anon):7> ((  0  ))\C-[[49m\C-[[K\C-M\n+(anon):62> l+zstyle:2> [[ 4 == 3 ]]C-[[49m\C-[[K\C-M\n+(anon):63> -z4h-get-cursor-pos\C-[[49m\C-[[K\C-M\n+-z4h-get-cursor-pos:3> local esc\C-[[49m\C-[[K+zstyle:5> zstyle -a :z4h:fzf-completion fzf-flags flagsr-pos:4> read -s -d R $\'esc?\\C-[[6n\' cursor_y cursor_x\C-[[49m\C-[[K\C-M\n+(anon)+-z4h-fzf:9> flags=( ) restore_screen  ))\C-[[49m\C-[[K\C-M\n+(anon):66> opts+=( --no-clear )\C-[[49m\C-[[K\C-M\n+(anon):67> ((  height <= c+-z4h-fzf:10> zstyle -a :z4h:fzf-completion fzf-bindings bindingsompt-at-bottom\C-[[49m\C-[[K\C-M\n+zstyle:2> [[ 3 == 3 && prompt-at-bottom +zstyle:2> [[ 4 == 3 ]]49m\C-[[K\C-M\n+zstyle:5> zstyle -T :z4h: prompt-at-bottom\C-[[49m\C-[[K\C-M\n+(anon):69> ((  height > LINES - cursor+zstyle:5> zstyle -a :z4h:fzf-completion fzf-bindings bindings 1  ))\C-[[49m\C-[[K\C-M\n+(anon):71> local move=$\'\\C-[[0m\\C-[[1;1H\'\C-[[4+-z4h-fzf:10> bindings=( ) pts+=( \'--layout=default\' )\C-[[49m\C-[[K\C-M\n+(anon):84> local _z4h_saved_screen\C-[[49m\C-[[K\C-M\n+(anon):8+-z4h-fzf:12> local -A keys=( ctrl-h backward-kill-word alt-j clear-query ctrl-u clear-query ctrl-k kill-line alt-k unix-line-discard ctrl-space toggle-out ctrl-a toggle-all )etopt n\C-[[49m\C-[[K\C-M\no_prompt_bang no_bg_nice no_aliases\'\C-[[49m\C-[[K\C-M\n+(eval):1> emulate -L+-z4h-fzf:22> ((  13  ))+(eval):2> setopt typeset_silent pipe_fail extended_glob prompt_percent no_prompt_subst\C-[[49m\C-[[K\C-M\n+(eval):3+-z4h-fzf:23> keys+=( tab up btab down ctrl-r up ctrl-s down ) \n+-z4h-save-screen:5> [[ -n /tmp/z4h-tmux-502-xterm-256color-1622236387.4660+-z4h-fzf:38> local kv49m\C-[[K\C-M\n+-z4h-save-screen:6> local pane\C-[[49m\C-[[K\C-M\n+-z4h-save-screen:7> pane=+-z4h-save-screen:7> TMUX=+-z4h-fzf:45> local k v-256color-1622236387.4660720825,3143,10 TMUX_PANE=\'\' /Users/hchen\C-[[49m\C-[[K\C-M\ng/.cache/zsh4humans/v5/tmux/bi+-z4h-fzf:46> local -i expect[49m\C-[[K\C-M\n\C-[[49m\C-[[K'
+-z4h-fzf:47> k=ctrl-spaceght=29' )
+-z4h-fzf:47> v=toggle-out
+-z4h-fzf:48> [[ +toggle-out+ == *+repeat+* ]]ail
+-z4h-fzf:47> k=btab
+-z4h-fzf:47> v=down
+-z4h-fzf:48> [[ +down+ == *+repeat+* ]]
+-z4h-fzf:47> k=ctrl-h
+-z4h-fzf:47> v=backward-kill-word
+-z4h-fzf:48> [[ +backward-kill-word+ == *+repeat+* ]]
+-z4h-fzf:47> k=ctrl-k
+-z4h-fzf:47> v=kill-line
+-z4h-fzf:48> [[ +kill-line+ == *+repeat+* ]]
+-z4h-fzf:47> k=ctrl-r
+-z4h-fzf:47> v=up
+-z4h-fzf:48> [[ +up+ == *+repeat+* ]]
+-z4h-fzf:47> k=ctrl-a
+-z4h-fzf:47> v=toggle-all
+-z4h-fzf:48> [[ +toggle-all+ == *+repeat+* ]]
+-z4h-fzf:47> k=ctrl-s
+-z4h-fzf:47> v=down
+-z4h-fzf:48> [[ +down+ == *+repeat+* ]]
+-z4h-fzf:47> k=alt-j
+-z4h-fzf:47> v=clear-query
+-z4h-fzf:48> [[ +clear-query+ == *+repeat+* ]]
+-z4h-fzf:47> k=alt-k
+-z4h-fzf:47> v=unix-line-discard
+-z4h-fzf:48> [[ +unix-line-discard+ == *+repeat+* ]]
+-z4h-fzf:47> k=tab
+-z4h-fzf:47> v=up
+-z4h-fzf:48> [[ +up+ == *+repeat+* ]]
+-z4h-fzf:47> k=ctrl-u
+-z4h-fzf:47> v=clear-query
+-z4h-fzf:48> [[ +clear-query+ == *+repeat+* ]]
+-z4h-fzf:59> local keymap
+-z4h-fzf:60> printf -v keymap %s:%s, ctrl-space toggle-out btab down ctrl-h backward-kill-word ctrl-k kill-line ctrl-r up ctrl-a toggle-all ctrl-s down alt-j clear-query alt-k unix-line-discard tab up ctrl-u clear-query
+-z4h-fzf:61> /Users/hcheng/.cache/zsh4humans/v5/fzf/bin/fzf '--query=^ab' '--color=hl:201,hl+:201' '--with-nth=2' '--delimiter=\000' --ansi --exact --no-mouse '--tiebreak=length,begin,index' --multi --cycle '--border=horizontal' --no-clear '--layout=default' '--height=29' '--bind=ctrl-space:toggle-out,btab:down,ctrl-h:backward-kill-word,ctrl-k:kill-line,ctrl-r:up,ctrl-a:toggle-all,ctrl-s:down,alt-j:clear-query,alt-k:unix-line-discard,tab:up,ctrl-u:clear-query'
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────






















  abc/subabc
  abd
> abc
  3/3 (0)
> ^ab
=======

**Comment by atjshop** [comment]
typing 'ab<tab>' doesn't have the subdirectory, while 'cd ab<tab>' still has:

![Screen Shot 2021-06-01 at 10 58 49 AM](https://user-images.githubusercontent.com/14037447/120355924-609c4080-c2c9-11eb-8304-a14617b04e82.jpg)
![Screen Shot 2021-06-01 at 10 59 18 AM](https://user-images.githubusercontent.com/14037447/120355940-65f98b00-c2c9-11eb-9c43-513986d77aab.jpg)

**Comment by romkatv** [comment]
Interesting. The body of `-z4h-comp-files` doesn't match its source code. Please post the output of the following command:

```zsh
autoload +X -Uz -- -z4h-comp-files; print -r -- $functions_source[-z4h-comp-files] && ls -l -- $Z4H/zsh4humans/fn/-z4h-comp-files*
```

**Comment by atjshop** [comment]
ah, did my installation messed up? below is the output:

/Users/xxxxxxx/.cache/zsh4humans/v5/zsh4humans/fn/-z4h-comp-files
16 -rw-r--r--  1 xxxxxxx  staff   4.1K May 28 15:54 /Users/xxxxxxx/.cache/zsh4humans/v5/zsh4humans/fn/-z4h-comp-files
24 -r--r--r--  1 xxxxxxx  staff    11K May 28 15:54 /Users/xxxxxxx/.cache/zsh4humans/v5/zsh4humans/fn/-z4h-comp-files.zwc

**Comment by romkatv** [comment]
That looks fine. Could you run `exec zsh` just to make sure you aren't using a shell that you've created before `z4h update`? Verify that after `exec zsh` you still have this issue. If you do, please attach `-z4h-comp-files` and `-z4h-comp-files.zwc` (see above for complete paths).

**Comment by atjshop** [comment]
Yeah shell was restarted after update. I just restarted shell again, still same.  Files attached.

[Archive.zip](https://github.com/romkatv/zsh4humans/files/6585310/Archive.zip)

**Comment by romkatv** [comment]
These files are fine. I think the output of `functions -- -z4h-comp-files` in https://github.com/romkatv/zsh4humans/issues/115#issuecomment-852193732 is incomplete. If you were to redirect it to a file, I think the file would have the same source as you have on disk.

Looking at the trace once again, it appears that you haven't set the style. Please copy paste this command verbatim into the shell and execute it:

```zsh
zstyle ':z4h:fzf-complete' recurse-dirs 'no'
```

Then try completing `cd` *within the same shell*.

**Comment by atjshop** [comment]
The style set was there in .zshrc file (As I stated in the first post and [here](https://github.com/romkatv/zsh4humans/issues/115#issuecomment-850689934) too.

I just verified it is in line 29 of ~/.zshrc. Does it matter the position it is in the file?
❯ rg 'recurse-dirs' ~/.zshrc
29:zstyle ':z4h:fzf-complete' recurse-dirs 'no'

I ran that line in shell directly and tried the 'cd ab<tab>', the subdirectory still appears.
